javaagent-tooling/javaagent-tooling-java9/src/main/java/io/opentelemetry/javaagent/tooling/SunMiscUnsafeGenerator.java

```diff
@@ -0,0 +1,399 @@
+/*
+ * Copyright The OpenTelemetry Authors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package io.opentelemetry.javaagent.tooling;
+
+import static net.bytebuddy.jar.asm.Opcodes.ACC_FINAL;
+import static net.bytebuddy.jar.asm.Opcodes.ACC_PRIVATE;
+import static net.bytebuddy.jar.asm.Opcodes.ACC_PUBLIC;
+import static net.bytebuddy.jar.asm.Opcodes.ACC_STATIC;
+import static net.bytebuddy.jar.asm.Opcodes.ACC_SUPER;
+import static net.bytebuddy.jar.asm.Opcodes.ALOAD;
+import static net.bytebuddy.jar.asm.Opcodes.DUP;
+import static net.bytebuddy.jar.asm.Opcodes.GETSTATIC;
+import static net.bytebuddy.jar.asm.Opcodes.ILOAD;
+import static net.bytebuddy.jar.asm.Opcodes.INVOKESPECIAL;
+import static net.bytebuddy.jar.asm.Opcodes.INVOKESTATIC;
+import static net.bytebuddy.jar.asm.Opcodes.INVOKEVIRTUAL;
+import static net.bytebuddy.jar.asm.Opcodes.IRETURN;
+import static net.bytebuddy.jar.asm.Opcodes.NEW;
+import static net.bytebuddy.jar.asm.Opcodes.PUTSTATIC;
+import static net.bytebuddy.jar.asm.Opcodes.RETURN;
+
+import io.opentelemetry.javaagent.bootstrap.AgentClassLoader;
+import java.lang.reflect.Field;
+import java.lang.reflect.Method;
+import java.lang.reflect.Modifier;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.List;
+import net.bytebuddy.jar.asm.ClassVisitor;
+import net.bytebuddy.jar.asm.ClassWriter;
+import net.bytebuddy.jar.asm.FieldVisitor;
+import net.bytebuddy.jar.asm.MethodVisitor;
+import net.bytebuddy.jar.asm.Opcodes;
+import net.bytebuddy.jar.asm.Type;
+
+/**
+ * Helper class for generating our own copy of sun.misc.Unsafe that delegates to
```

Comment:
Just curious if you considered an approach that just compiles Java, I guess we'd be able to expose `jdk.internal` via `--add-exports`?

Reply:
this sounds nice if it can work

Reply:
My initial attempt was to generate a java version of sun.misc.Unsafe that uses method handles (I went for method handles from the start because I had a vague memory that some methods in internal unsafe have been renamed) to forward to internal unsafe but this failed with `error: package exists in another module: jdk.unsupported`. I did try a bit to coax it to work but failed. In retrospect I think I should have just moved it into a module compiled with java8, though then if I'd want to replace some method handles with direct calls I'd probably need to create a stub of internal unsafe. Anyway while fighting the compiler I realized that I could just generate bytecode and asm is my favourite framework and I haven't had a chance to use it for almost a year so it was pretty much drop everything else and start generating bytecode.